### PR TITLE
Move server implementation out of server program for testability

### PR DESCRIPTION
**Pull request by nguerrera**
Contributes to https://github.com/Azure/cadl-azure/issues/615

* Refactor things so that a language server can be tested without starting a program.
* Port existing completion tests over to new mechanism.
  * This revealed two small bugs, which are fixed here:
    * Symbols pulled in via using were given the function icon. 
    * `true` and `false` were offered as completions after `.`.
 * More language server tests to follow in separate PRs that leverage this.
     
The diff as shown here will be hard to read because I moved 99% of server.ts to serverlib.ts. I had to rename the file because the runner that must be cross-version compatible expects server.js to be the program entry point. It would have been cleaner to create a new server-program.ts or something, but I didn't think it was worth breaking earlier runners for that. So, to help review, I made a fake commit here that shows the meaningful changes to the language server code:

https://github.com/nguerrera/cadl/pull/1/files?diff=unified&w=1

### CODE REVIEW

File: packages/compiler/test/test-host.ts
Comment (timotheeguerin) [inline packages/compiler/test/test-host.ts:45]: what was the reasoning behind that change? then entire system should have already been os agnostic so shouldn't really have to do that

Comment (nguerrera) [reply packages/compiler/test/test-host.ts:45]: The immediate reason was we're calling pathToFileURL and it throws on windows on "/something". Unfortunately, assuming "/" is an absolute path is not really "OS-agnostic". It's a current-drive relative path on Windows.

Comment (nguerrera) [reply packages/compiler/test/test-host.ts:45]: Hmm, that's not what's throwing, that resolves to file://C:/something. I need to revert and check what the error was. On it.

File: packages/compiler/server/serverlib.ts
Comment (timotheeguerin) [inline packages/compiler/server/serverlib.ts:686]: There was a bug here, I fixed it in the emitter PR but maybe this could be the good place to do so as well https://github.com/microsoft/cadl/pull/41/files#diff-97b4206fa6eac301190b0164548985edc255b40970d3a46961e9a9cb83a41f83R687

File: packages/compiler/test/server/completion.ts
Comment (timotheeguerin) [inline packages/compiler/test/server/completion.ts:146]: Does this one work fine? I remember it wasn't able to figure out before if the cursor was in the extends section or inside the model body so couldn't complete other model as it would confuse when trying to write properties

**Review by timotheeguerin — APPROVED** [review]

File: packages/compiler/test/test-host.ts
Comment (timotheeguerin) [reply packages/compiler/test/test-host.ts:45]: I see yeah seems like that's the only way if we have to use `pathToFileURL`. However is the one causing issue the one defined in the test host? in that case couldn't we just say instead that the url is `file://${path}`

File: packages/compiler/test/server/completion.ts
Comment (nguerrera) [reply packages/compiler/test/server/completion.ts:146]: It does work:

image

We don't need to know that we're in the extends section for this test, just that we're in a type reference. And I think we can actually find out we're in the heritage here if we needed to for some reason.  But I think the problem you're thinking about would be offering `extends` as a completion before the user types it or `{`? That is harder, still doesn't work, and not tested here.

Comment (timotheeguerin) [reply packages/compiler/test/server/completion.ts:146]: oh right that must have been the issue

File: packages/compiler/server/serverlib.ts
Comment (nguerrera) [reply packages/compiler/server/serverlib.ts:451]: No, it fixes an issue where we would get function icon for anything pulled in via using. Previously symbol was either type or decorator, but now it can be using and we need to get the symbol source to pick the right icon.

File: packages/compiler/test/test-host.ts
Comment (nguerrera) [reply packages/compiler/test/test-host.ts:45]: So, it was actually the opposite, I used file:/// as the root uri since the root of virtual file system was "/" and then URLToFilePath blew up. I think it's easier to just have the virtual root have a drive letter, but I'll add a comment to clarify why.

File: packages/compiler/server/server.ts
Comment (nguerrera) [reply packages/compiler/server/server.ts:12]: server could be undefined in fatalError if we throw before createServer. I fixed this using a local `s` that aliases `server`. If there's a prettier way to do this, please let me know. :)

**Comment by nguerrera** [comment]
I'll wait for some other big PRs to merge before merging this to make sure all changes to server.ts go to new serverlib.ts without putting that on others. 😫😁

**Pull request merged**
